File: CCBTS05.py
'''

* Author : Ayushman Chahar #
* About  : IT Sophomore    #
* Insti  : VIT, Vellore    #

'''

import os
import sys
# from collections import *
# from itertools import *
# from math import *
# from queue import *
# from heapq import *
# from bisect import *
from io import BytesIO, IOBase

BUFSIZE = 8192


class FastIO(IOBase):
    newlines = 0

    def __init__(self, file):
        self._fd = file.fileno()
        self.buffer = BytesIO()
        self.writable = "x" in file.mode or "r" not in file.mode
        self.write = self.buffer.write if self.writable else None

    def read(self):
        while True:
            b = os.read(self._fd, max(os.fstat(self._fd).st_size, BUFSIZE))
            if not b:
                break
            ptr = self.buffer.tell()
            self.buffer.seek(0, 2), self.buffer.write(b), self.buffer.seek(ptr)
        self.newlines = 0
        return self.buffer.read()

    def readline(self):
        while self.newlines == 0:
            b = os.read(self._fd, max(os.fstat(self._fd).st_size, BUFSIZE))
            self.newlines = b.count(b"\n") + (not b)
            ptr = self.buffer.tell()
            self.buffer.seek(0, 2), self.buffer.write(b), self.buffer.seek(ptr)
        self.newlines -= 1
        return self.buffer.readline()

    def flush(self):
        if self.writable:
            os.write(self._fd, self.buffer.getvalue())
            self.buffer.truncate(0), self.buffer.seek(0)


class IOWrapper(IOBase):
    def __init__(self, file):
        self.buffer = FastIO(file)
        self.flush = self.buffer.flush
        self.writable = self.buffer.writable
        self.write = lambda s: self.buffer.write(s.encode("ascii"))
        self.read = lambda: self.buffer.read().decode("ascii")
        self.readline = lambda: self.buffer.readline().decode("ascii")


sys.stdin, sys.stdout = IOWrapper(sys.stdin), IOWrapper(sys.stdout)
input = lambda: sys.stdin.readline().rstrip("\r\n")
readint = lambda: int(sys.stdin.readline().rstrip("\r\n"))
readints = lambda: map(int, sys.stdin.readline().rstrip("\r\n").split())
readstr = lambda: sys.stdin.readline().rstrip("\r\n")
readstrs = lambda: map(str, sys.stdin.readline().rstrip("\r\n").split())
readarri = lambda: [int(_) for _ in sys.stdin.readline().rstrip("\r\n").split()]
readarrs = lambda: [str(_) for _ in sys.stdin.readline().rstrip("\r\n").split()]

mod = 998244353
MOD = int(1e9) + 7

N = int(1e5) + 1
sys.setrecursionlimit(N)
distances = [0] * N
height, euler, first, segtree, visited = [], [], [], [], []
n = None


def clear_global():
    global height, euler, first, segtree, visited, n
    height.clear()
    euler.clear()
    first.clear()
    segtree.clear()
    visited.clear()


def dfs(adj, node, h=0):
    global visited, height, euler, distances
    visited[node] = 1
    height[node] = h
    first[node] = len(euler)
    euler.append(node)
    for i in adj[node]:
        if(not (visited[i[0]])):
            distances[i[0]] = distances[node] + i[1]
            dfs(adj, i[0], h + 1)
            euler.append(node)


def build(node, b, e):
    global segtree, euler, height
    if (b == e): segtree[node] = euler[b]
    else:
        mid = (b + e) >> 1
        build(node << 1, b, mid)
        build(node << 1 | 1, mid + 1, e)
        l = segtree[node << 1]
        r = segtree[node << 1 | 1]
        segtree[node] = l if (height[l] < height[r]) else r


def query(node, b, e, L, R):
    global segtree, height
    if (b > R or e < L): return -1
    if (b >= L and e <= R): return segtree[node]
    mid = (b + e) >> 1
    left = query(node << 1, b, mid, L, R)
    right = query(node << 1 | 1, mid + 1, e, L, R)
    if (left == -1): return right
    if (right == -1): return left
    return left if (height[left] < height[right]) else right


def preC(adj, root=0):
    global height, euler, first, segtree, visited, n
    n = len(adj)
    height = [0 for _ in range(n)]
    first = [0 for _ in range(n)]
    visited = [0 for _ in range(n)]
    dfs(adj, root)
    m = len(euler)
    segtree = [0 for _ in range(m * 4)]
    build(1, 0, m - 1)


def lca(u, v):
    global first, euler
    left, right = first[u], first[v]
    right = first[v]
    if (left > right): left, right = right, left
    return query(1, 0, len(euler) - 1, left, right)


def solve():
    global distances
    n, q, R = readints()
    adj = [[] for _ in range(n)]
    distances = [0 for _ in range(n + 1)]
    for _ in range(n - 1):
        l, r, dist = readints()
        adj[l - 1].append([r - 1, dist])
        adj[r - 1].append([l - 1, dist])
    preC(adj)
    for _ in range(q):
        u, v = readints()
        u -= 1; v -= 1
        LCA = lca(u, v)
        print(distances[u] + distances[v] - distances[LCA])
    clear_global()


def main():
    t = 1
    t = readint()
    for _ in range(t):
        # print("Case #" + str(_ + 1) + ": ", end="")
        solve()


if __name__ == "__main__":
    main()

File: CCBTS05.cpp
/*

*  Author : Ayushman Chahar   #
*  About  : IT Sophomore      #
*  Insti  : VIT, Vellore      #

*/

#pragma GCC optimize("Ofast")  
#pragma GCC target("avx,avx2,fma") 
#pragma GCC optimization ("unroll-loops")

#include<bits/stdc++.h>
using namespace std;

#define MOD 1000000007
#define mod 998244353

typedef long long int ll;
typedef unsigned long long int ull;

template<typename T, typename U> inline void amin(T &x, U y) {if(y < x) x = y;}
template<typename T, typename U> inline void amax(T &x, U y) {if(x < y) x = y;}

const int N = 1e5 + 2;
ll distances[N];

struct LCA {
    vector<int> height, euler, first, segtree;
    vector<bool> visited;
    int n;

    LCA(vector<vector<pair<int, int>>> &adj, int root) {
        n = adj.size();
        height.resize(n);
        first.resize(n);
        euler.reserve(n * 2);
        visited.assign(n, false);
        dfs(adj, root);
        int m = euler.size();
        segtree.resize(m * 4);
        build(1, 0, m - 1);
    }

    void dfs(vector<vector<pair<int, int>>> &adj, int node, int h = 0) {
        visited[node] = true;
        height[node] = h;
        first[node] = euler.size();
        euler.push_back(node);
        for (auto sec : adj[node]) {
            if (!visited[sec.first]) {
                distances[sec.first] = distances[node] + sec.second;
                dfs(adj, sec.first, h + 1);
                euler.push_back(node);
            }
        }
    }

    void build(int node, int b, int e) {
        if (b == e) {
            segtree[node] = euler[b];
        } else {
            int mid = (b + e) / 2;
            build(node << 1, b, mid);
            build(node << 1 | 1, mid + 1, e);
            int l = segtree[node << 1], r = segtree[node << 1 | 1];
            segtree[node] = (height[l] < height[r]) ? l : r;
        }
    }

    int query(int node, int b, int e, int L, int R) {
        if (b > R || e < L)
            return -1;
        if (b >= L && e <= R)
            return segtree[node];
        int mid = (b + e) >> 1;

        int left = query(node << 1, b, mid, L, R);
        int right = query(node << 1 | 1, mid + 1, e, L, R);
        if (left == -1) return right;
        if (right == -1) return left;
        return height[left] < height[right] ? left : right;
    }
    int lca(int u, int v) {
        int left = first[u], right = first[v];
        if (left > right)
            swap(left, right);
        return query(1, 0, euler.size() - 1, left, right);
    }
};

void solve() {
    int n, q, R;
    cin >> n >> q >> R;
    vector<vector<pair<int, int>>> adj;
    adj.resize(n);
    for(int i = 0; i < n - 1; i++) {
        int l, r, dist;
        cin >> l >> r >> dist;
        adj[l - 1].emplace_back(make_pair(r - 1, dist));
        adj[r - 1].emplace_back(make_pair(l - 1, dist));
    }
    LCA obj(adj, R - 1);
    for(int i = 0; i < q; i++) {
        int u, v;
        cin >> u >> v;
        int myLCA = obj.lca(u - 1, v - 1);
        cout << distances[u - 1] + distances[v - 1] - 2 * distances[myLCA] << "\n";
    }
}

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(nullptr); cout.tie(nullptr);
    int t = 1, casee = 1;
    cin >> t;
    while(t--) {
        // cout << "Case #" << casee++ << ": ";
        solve();
    }
    return 0;
}